R/utilities-units.R

```diff
@@ -336,3 +336,200 @@ initializeDimensionAndUnitLists <- function() {
+  # containing source units remain untouched.
+  #
+  # These newly created columns are removed before the converted data frame is
+  # returned to the user.
+  data <- data %>%
+    dplyr::mutate(
+      dplyr::across(
+        .cols = dplyr::matches("Unit$|Weight$"),
+        .fns = as.character,
+        .names = "{.col}Split"
+      )
+    ) %>%
+    dplyr::mutate(
+      dplyr::across(
+        .cols = dplyr::matches("Split$"),
+        .fns = ~ tidyr::replace_na(.x, "missing")
+      )
+    ) %>%
+    # `split()` will change the row order of the data frame depending on the
+    # alphabetical order of the levels of the variable used to split the data
+    # frame into a list.
+    #
+    # Therefore, an internal row identifier is kept to restore the original
+    # data frame row order before data is returned.
+    dplyr::mutate(.rowidInternal = dplyr::row_number())
+
+  # xUnit --------------------------
+
+  xDataList <- .removeEmptyDataFrame(split(data, data$xUnitSplit))
+  data <- dplyr::bind_rows(lapply(xDataList, .xUnitConverter, xTargetUnit))
+
+  # yUnit ----------------
+
+  yDataList <- .removeEmptyDataFrame(split(data, list(data$yUnitSplit, data$molWeightSplit)))
+  data <- dplyr::bind_rows(lapply(yDataList, .yUnitConverter, yTargetUnit))
+
+  # yUnit error ----------------
+
+  if ("yErrorValues" %in% names(data)) {
+    yErrorDataList <- .removeEmptyDataFrame(split(data, list(data$yErrorUnitSplit, data$molWeightSplit)))
+    data <- dplyr::bind_rows(lapply(yErrorDataList, .yErrorUnitConverter, yTargetUnit))
+  }
+
+  # Restore the original row order using the internal row id
+  data <- dplyr::arrange(data, .rowidInternal)
+
+  # Remove all columns that were added only for internal workings of the function.
+  data <- dplyr::select(data, -dplyr::matches("Split$|.rowidInternal"))
+
+  return(data)
+}
+
+#' Remove empty data frames sometimes produced due to the non-existent
+#' combination of source unit and molecular unit.
+#'
+#' @keywords internal
+#' @noRd
+.removeEmptyDataFrame <- function(x) {
+  x <- purrr::keep(x, ~ nrow(.x) > 0L)
+  x
+}
+
+#' @keywords internal
+#' @noRd
+.xUnitConverter <- function(xData, xTargetUnit) {
+  xData$xValues <- ospsuite::toUnit(
+    quantityOrDimension = xData$xDimension[[1]],
+    values = xData$xValues,
+    targetUnit = xTargetUnit,
+    sourceUnit = xData$xUnit[[1]]
+  )
+
+  xData$xUnit <- xTargetUnit
+
+  return(xData)
+}
+
+#' @keywords internal
+#' @noRd
+.yUnitConverter <- function(yData, yTargetUnit) {
+  yData$yValues <- toUnit(
+    quantityOrDimension = yData$yDimension[[1]],
+    values = yData$yValues,
+    targetUnit = yTargetUnit,
+    sourceUnit = yData$yUnit[[1]],
+    molWeight = yData$molWeight[[1]],
+    molWeightUnit = ospUnits$`Molecular weight`$`g/mol`
+  )
+
+  yData$yUnit <- yTargetUnit
+
+  return(yData)
+}
+
+#' @keywords internal
+#' @noRd
+.yErrorUnitConverter <- function(yData, yTargetUnit) {
```

Comment:
In general, method with side effects should not return anything, unless this is used for method chaining. Is this the case here?

Reply:
I cannot tell from the way it's used 
`data <- dplyr::bind_rows(lapply(yDataList, .yUnitConverter, yTargetUnit))`

What does `lapply(yDataList, .yUnitConverter, yTargetUnit)` means
Apply yUnitConverter to all element in yData and pass yTargetUnit as second parameter?

Does yUnitConverter  needs to return something?

Reply:
> Apply yUnitConverter to all element in yData and pass yTargetUnit as second parameter?

Exactly. 

I'd like to use `purrr::map(yDataList, ~ .yUnitConverter(.x, yTargetUnit))` which makes this clearer, but I've been told to use as much base-R as possible.

> Does .yUnitConverter needs to return something?

Yes, for each data frame in the list, ` .yUnitConverter()` returns a data frame with converted units. These data frames are then bound by rows by  `dplyr::bind_rows()`.

Reply:
I Think `~ .yUnitConverter(.x, yTargetUnit)` is not clear at all if you are not familiar with the package
I do not know what `.x` means or what `~` means. This is probably very specific to purr. (or maybe not)

I think either code is fine. We are now using 3rd party packages anwyays so we might as well us it as efficient as possible. What we are missing, is a short summary of the main functions of dplyr and purr. I cannot take the time to navigate to their doc and try to understand what this means when there is so much to such for Instead we need a dpply and purr for dummy wiki where we explain the methods we are using with concrete examples